### PR TITLE
feat(fortran): Fortran interface over the C FFI

### PR DESCRIPTION
**Pull request by tlmquintino**
## Summary

A shallow, idiomatic **Fortran 2018** interface (`fortran/`) so scientific Fortran codes (IFS and friends) can read and write Tensogram messages directly, binding straight to the existing `libtensogram` C ABI — no new C or Rust code.

Full design + staged roadmap: [`PLAN_FORTRAN.md`](https://github.com/ecmwf/tensogram/blob/feat/fortran-interface/PLAN_FORTRAN.md).

## Approach (Path A)

- Hand-written `iso_c_binding` module over the existing C ABI (the `eccodes_f90` model IFS users expect).
- Native `real(:,:)` ergonomics via the F2008 `contiguous` attribute — the Fortran array descriptor never crosses the boundary, so no `CFI_cdesc_t`/Rust work is needed.

## Delivered surface (synchronous)

- **Generic encode/decode** over dtype (real32/64, int32/64) and rank (0–7).
- **File API**: multi-message `.tgm` open/create/append/random-access decode/read.
- **Application metadata** (zero-dependency `tensogram_meta` builder + dot-notation getters) and the **encoding pipeline** (optional encoding/filter/compression, e.g. lossless zstd).
- **Streaming encoder** (progressive multi-object write).
- RAII non-copyable handles with an `error stop` copy guard.

Only the **async surface is deferred** (PLAN_FORTRAN.md §7).

## Column-major contract

A Fortran `a(ni,nj)` is written with the on-wire shape/strides **reversed** to C order: Fortran↔Fortran round-trips are bit-identical; NumPy/C readers see the transpose.

## Verification

- 24 CTest tests; module compiles clean under `-std=f2018 -Wall -Wextra -Werror`; **95.7%** line coverage.
- **Bidirectional cross-language parity in CI**: Fortran↔C/C++ and Fortran↔Python (both directions), through a lossless pipeline.
- Error-enum↔`tensogram.h` consistency guard; edge-case suite; example smoke tests.
- CMake (system of record) + fpm (drift-checked); a `Fortran (Linux)` CI lane.

Distribution via `fpm` + CMake/pkg-config against the shipped `tensogram.pc`.

Docs Preview
https://sites.ecmwf.int/docs/tensogram/pull-requests/PR-127

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -475,6 +475,54 @@ jobs:
...
+
+      - name: Install the tensogram C ABI (cargo-c) and expose via pkg-config
+        run: |
+          PFX="$GITHUB_WORKSPACE/.ffi-prefix"
+          cargo cinstall --release -p tensogram-ffi --prefix="$PFX" --libdir=lib
+          echo "PKG_CONFIG_PATH=$PFX/lib/pkgconfig" >> "$GITHUB_ENV"
+          echo "LD_LIBRARY_PATH=$PFX/lib" >> "$GITHUB_ENV"
+
+      - name: Build + test the Fortran binding (CMake + CTest)
+        run: make fortran-test
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply .github/workflows/ci.yml:568]: Fixed in 79a7955 — the Fortran CI job now installs fpm and runs `make fortran-fpm-test` as a drift check against the authoritative CMake build.

[the remark this answers is not in the document]
Comment (tlmquintino) [reply .github/workflows/ci.yml:568]: Fixed in 79a7955 — the job now builds the Python binding via `maturin develop` into a venv placed first on PATH, so CMake discovers it and the Fortran↔Python parity tests run in CI.

File: PLAN_FORTRAN.md [hunk 2]
```diff
@@ -0,0 +1,1369 @@
...
+   user never hand-rolls `c_loc`, strides, or byte counts.
+
+Deferred (documented, not built in the first arc):
+
+- The **streaming encoder** (progressive writes).
+- The entire **async** surface (it uses C completion callbacks — the
+  hardest part to bind and the least in demand for NWP codes, which are
+  happy with blocking calls).
+- `CFI_cdesc_t` / TS 29113 descriptor passing (only needed for
+  zero-copy *non-contiguous* input — see §5.2).
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply PLAN_FORTRAN.md]: Fixed in 1edc724 — removed the stale streaming entry from the §1 Deferred list; §7 marks the streaming encoder as delivered.

File: PLAN_FORTRAN.md [hunk 3]
```diff
@@ -0,0 +1,1369 @@
+# Tensogram Fortran Interface — Development Plan
+
+> **Status.** The **synchronous surface is implemented** on branch
+> `feat/fortran-interface` (PR #127): generic encode/decode (dtype × rank),
+> the multi-message file API, application metadata + the encoding pipeline,
+> and the streaming encoder — with bidirectional cross-language parity
+> (C/C++ and Python), an error-enum↔header consistency guard, and a CI
+> lane. Only the **async surface is deferred** (§7). This document remains
+> the durable design + roadmap; per-milestone status is marked in §7 and
+> the backward-looking record of merged work is `../CHANGELOG.md`.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply PLAN_FORTRAN.md]: Fixed in 1edc724 — changed `../CHANGELOG.md` to `CHANGELOG.md` (PLAN_FORTRAN.md is at the repository root).

File: PLAN_FORTRAN.md [hunk 4]
```diff
@@ -0,0 +1,1369 @@
...
+---
+
+## 7. Development roadmap (named, ordered milestones)
+
+Each milestone is independently shippable, lists its acceptance criteria,
+and — per AGENTS.md — leads with how it is **tested/verified** (TDD,
+behaviour-driven). Ordering is by dependency, not by ceremony.
+
+> **Delivery status (✓ = merged on `feat/fortran-interface`; see
+> `../CHANGELOG.md` for specifics).** The original deliverable text below
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply PLAN_FORTRAN.md]: Fixed in 1edc724 — same root-relative link corrected to `CHANGELOG.md`.

File: fortran/src/tensogram.F90 [hunk 5]
```diff
@@ -0,0 +1,1427 @@
...
+   end subroutine buffer_final
+
+   subroutine buffer_assign(lhs, rhs)
+      class(tensogram_buffer), intent(out) :: lhs
+      type(tensogram_buffer),  intent(in)  :: rhs
+      if (c_associated(rhs%raw%data)) then
+         error stop "tensogram_buffer is non-copyable: assigning a live buffer would alias and double-free; pass by reference"
+      else
+         error stop "tensogram_buffer is non-copyable: do not assign buffers; pass by reference"
+      end if
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.F90:735]: False positive. The module compiles clean under `-std=f2018 -Wall -Wextra -Werror` (verified locally and by the CI Fortran job, which uses the same flags). gfortran does not emit `-Wunused-dummy-argument` for an `intent(out)` derived-type dummy that has default initialisation: `lhs` is default-initialised (`ptr = c_null_ptr`) on entry, so it is treated as set, not unused. (The earlier genuinely-unused `rhs` was a real warning and is already fixed via the `c_associated(rhs%...)` reference.)

File: fortran/src/tensogram.F90 [hunk 6]
```diff
@@ -0,0 +1,1427 @@
...
+   end subroutine message_final
+
+   subroutine message_assign(lhs, rhs)
+      class(tensogram_message), intent(out) :: lhs
+      type(tensogram_message),  intent(in)  :: rhs
+      if (c_associated(rhs%ptr)) then
+         error stop "tensogram_message is non-copyable: assigning a live handle would alias and double-free; pass by reference"
+      else
+         error stop "tensogram_message is non-copyable: do not assign handles; pass by reference"
+      end if
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.F90:762]: False positive. The module compiles clean under `-std=f2018 -Wall -Wextra -Werror` (verified locally and by the CI Fortran job, which uses the same flags). gfortran does not emit `-Wunused-dummy-argument` for an `intent(out)` derived-type dummy that has default initialisation: `lhs` is default-initialised (`ptr = c_null_ptr`) on entry, so it is treated as set, not unused. (The earlier genuinely-unused `rhs` was a real warning and is already fixed via the `c_associated(rhs%...)` reference.)

File: fortran/src/tensogram.F90 [hunk 7]
```diff
@@ -0,0 +1,1427 @@
...
+   end subroutine file_final
+
+   subroutine file_assign(lhs, rhs)
+      class(tensogram_file), intent(out) :: lhs
+      type(tensogram_file),  intent(in)  :: rhs
+      if (c_associated(rhs%ptr)) then
+         error stop "tensogram_file is non-copyable: assigning a live handle would alias and double-close; pass by reference"
+      else
+         error stop "tensogram_file is non-copyable: do not assign handles; pass by reference"
+      end if
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.F90:1000]: False positive. The module compiles clean under `-std=f2018 -Wall -Wextra -Werror` (verified locally and by the CI Fortran job, which uses the same flags). gfortran does not emit `-Wunused-dummy-argument` for an `intent(out)` derived-type dummy that has default initialisation: `lhs` is default-initialised (`ptr = c_null_ptr`) on entry, so it is treated as set, not unused. (The earlier genuinely-unused `rhs` was a real warning and is already fixed via the `c_associated(rhs%...)` reference.)

File: fortran/src/tensogram.F90 [hunk 8]
```diff
@@ -0,0 +1,1427 @@
...
+   end subroutine metadata_final
+
+   subroutine metadata_assign(lhs, rhs)
+      class(tensogram_metadata), intent(out) :: lhs
+      type(tensogram_metadata),  intent(in)  :: rhs
+      if (c_associated(rhs%ptr)) then
+         error stop "tensogram_metadata is non-copyable: assigning a live handle would alias and double-free; pass by reference"
+      else
+         error stop "tensogram_metadata is non-copyable: do not assign handles; pass by reference"
+      end if
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.F90:1248]: False positive. The module compiles clean under `-std=f2018 -Wall -Wextra -Werror` (verified locally and by the CI Fortran job, which uses the same flags). gfortran does not emit `-Wunused-dummy-argument` for an `intent(out)` derived-type dummy that has default initialisation: `lhs` is default-initialised (`ptr = c_null_ptr`) on entry, so it is treated as set, not unused. (The earlier genuinely-unused `rhs` was a real warning and is already fixed via the `c_associated(rhs%...)` reference.)

File: fortran/src/tensogram.F90 [hunk 9]
```diff
@@ -0,0 +1,1427 @@
...
+   end subroutine stream_enc_final
+
+   subroutine stream_enc_assign(lhs, rhs)
+      class(tensogram_streaming_encoder), intent(out) :: lhs
+      type(tensogram_streaming_encoder),  intent(in)  :: rhs
+      if (c_associated(rhs%ptr)) then
+         error stop "tensogram_streaming_encoder is non-copyable (live handle); pass by reference"
+      else
+         error stop "tensogram_streaming_encoder is non-copyable; pass by reference"
+      end if
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.F90:1318]: False positive. The module compiles clean under `-std=f2018 -Wall -Wextra -Werror` (verified locally and by the CI Fortran job, which uses the same flags). gfortran does not emit `-Wunused-dummy-argument` for an `intent(out)` derived-type dummy that has default initialisation: `lhs` is default-initialised (`ptr = c_null_ptr`) on entry, so it is treated as set, not unused. (The earlier genuinely-unused `rhs` was a real warning and is already fixed via the `c_associated(rhs%...)` reference.)

File: fortran/src/tensogram.F90 [hunk 10]
```diff
@@ -0,0 +1,1427 @@
...
+   subroutine buffer_final(self)
+      type(tensogram_buffer), intent(inout) :: self
+      call self%free()
+   end subroutine buffer_final
+
+   subroutine buffer_assign(lhs, rhs)
+      class(tensogram_buffer), intent(out) :: lhs
+      type(tensogram_buffer),  intent(in)  :: rhs
+      if (c_associated(rhs%raw%data)) then
+         error stop "tensogram_buffer is non-copyable: assigning a live buffer would alias and double-free; pass by reference"
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.F90:732]: False positive. Line 647 is 126 characters and the longest line in the file is 128 — both within the 132-column free-form limit, so `-Werror=line-truncation` does not fire (the module compiles clean under `-Wall -Werror`). The streaming-guard message was already shortened when it genuinely exceeded the limit.

File: Makefile [hunk 11]
```diff
@@ -129,6 +130,26 @@ cpp-build: ## Build C++ tests via CMake
...
+# CMake is the system of record; fpm is retained as a devloop. Not folded into
+# the default `make test` because it needs a Fortran compiler and an installed
+# tensogram.pc.
+
+fortran-build: ## Build the Fortran binding + examples + tests via CMake
+	cmake -S fortran -B build/fortran -DCMAKE_BUILD_TYPE=Debug
+	cmake --build build/fortran
+
+fortran-test: fortran-build ## Run Fortran tests (incl. Fortran<->Python parity when available)
+	ctest --test-dir build/fortran --output-on-failure
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply Makefile]: Fixed in afa8862 — changed to `cd build/fortran && ctest` so the target works with the ctest shipped by the declared CMake 3.16 minimum (`--test-dir` is 3.20+), matching the cpp-test pattern.

File: plans/TODO.md [hunk 12]
```diff
@@ -165,6 +165,37 @@ streaming encoder.  See the *Asynchronous C++ API* entry in
...
+      later option only if zero-copy *non-contiguous* input is ever
+      required.
+    - **Column-major contract**: a Fortran `a(ni,nj)` is written with the
+      on-wire descriptor shape/strides **reversed** to C order, so
+      Fortran↔Fortran round-trips are bit-identical and a NumPy / C
+      reader sees the transpose. This is the single most consequential
+      decision — document it loudly against the existing readers.
+    - **Synchronous first.** Ship encode/decode + the file API; defer the
+      streaming encoder and the whole async surface (the C async path
+      uses completion callbacks — hardest part, least demand).
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply plans/TODO.md]: Fixed in d23ecb4 — the TODO 'Synchronous first' bullet now lists the streaming encoder as shipped; only async is deferred.

File: docs/src/guide/fortran-api.md [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+```
+
+In your own project:
+
+```cmake
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(TENSOGRAM REQUIRED IMPORTED_TARGET tensogram)
+add_executable(my_app my_app.f90)
+target_link_libraries(my_app PRIVATE tensogram_f)   # + PkgConfig::TENSOGRAM
+```
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply docs/src/guide/fortran-api.md:74]: Fixed — the snippet no longer references the build-internal `tensogram_f` target; it compiles the single module file and links `PkgConfig::TENSOGRAM`, with `add_subdirectory(fortran)` shown as the alternative.

File: docs/src/guide/fortran-api.md [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+| `tensogram_object_dtype(msg, iobj)` | dtype string (e.g. `"float32"`) |
+| `tensogram_to_array(msg, iobj, out, err)` | Copy an object into a Fortran array — **generic** over dtype and rank of `out` |
+| `tensogram_file_open/create(path, file, err)` | Open / create a `.tgm` file |
+| `tensogram_file_message_count(file, n, err)` | Number of messages in the file |
+| `tensogram_file_append(file, a, err [, metadata_json, hash, encoding, filter, compression])` | Encode `a` and append it as a message — **generic** over dtype and rank |
+| `tensogram_file_decode_message(file, index, msg, err [, verify_hash, native_byte_order])` | Decode message `index` (1-based) |
+| `tensogram_file_read_message(file, index, buf, err)` | Read raw message bytes at `index` |
+| `tensogram_meta` + `%add_string` / `%add_int` / `%add_real` / `%base_json` | Build per-object application metadata (JSON-escaped) |
+| `tensogram_message_metadata(msg, meta, err)` | Extract a metadata handle from a decoded message |
+| `tensogram_metadata_get_string/int/float(meta, key [, default])` | Look up metadata by dot-notation key |
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply docs/src/guide/fortran-api.md]: Fixed — split the row: `_get_string(meta, key)` takes no default (returns empty when absent), while `_get_int` / `_get_float` require `default`.

File: PLAN_FORTRAN.md [hunk 15]
```diff
@@ -0,0 +1,1369 @@
+# Tensogram Fortran Interface — Development Plan
+
+> **Status.** The **synchronous surface is implemented** on branch
+> `feat/fortran-interface` (PR #127): generic encode/decode (dtype × rank),
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply PLAN_FORTRAN.md]: Fixed — removed the branch/PR reference from the status banner per AGENTS.md; it now describes the delivered surface without naming the branch.

File: PLAN_FORTRAN.md [hunk 16]
```diff
@@ -0,0 +1,1369 @@
...
+## 7. Development roadmap (named, ordered milestones)
+
+Each milestone is independently shippable, lists its acceptance criteria,
+and — per AGENTS.md — leads with how it is **tested/verified** (TDD,
+behaviour-driven). Ordering is by dependency, not by ceremony.
+
+> **Delivery status (✓ = merged on `feat/fortran-interface`; see
+> `CHANGELOG.md` for specifics).** The original deliverable text below
+> is preserved as the roadmap; deviations from what actually shipped are
+> noted inline.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply PLAN_FORTRAN.md]: Fixed — same process-ephemeral branch reference removed from the §7 delivery-status note.

File: fortran/src/tensogram.f90 [hunk 17]
```diff
@@ -0,0 +1,1427 @@
...
+      character(len=*),   intent(in)           :: dtype
+      character(len=*),   intent(in), optional :: extra
+      character(len=*),   intent(in), optional :: encoding, filter, compression
+      character(len=:), allocatable :: js
+      js = '{"descriptors":[' // &
+           descriptor_object_json(fshape, dtype, encoding, filter, compression) // ']'
+      if (present(extra)) then
+         if (len_trim(extra) > 0) js = js // ',' // trim(extra)
+      end if
+      js = js // '}'
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.f90]: Fixed in c40e5da — the descriptor builder now strips the outer braces of a complete JSON object, so encode/append accept EITHER a complete object (`{}`, `{"base":[...]}` — same shape as the streaming `metadata_json`) OR the builder's fragment (`tensogram_meta%base_json()`). Added a complete-object round-trip test and documented both forms.

File: CHANGELOG.md [hunk 18]
```diff
@@ -5,6 +5,63 @@ Format follows [Keep a Changelog](https://keepachangelog.com/).
...
+`tensogram_message_metadata` + the dot-notation getters
+`tensogram_metadata_get_string` / `_int` / `_float`.
+
+**Streaming encoder.** `tensogram_streaming_encoder` writes a single
+multi-object message to a file progressively — preamble + header metadata
+on `_create`, one data-object frame per `_write` (generic over dtype and
+rank, with optional compression), footer + postamble on `_finish` —
+without buffering the whole message. Non-copyable RAII handle;
+`tensogram_streaming_encoder_count` reports progress. New
+`examples/fortran/streaming.f90`.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply CHANGELOG.md:49]: Fixed — updated the PR description: streaming is listed under the delivered surface, with only the async surface deferred.

File: fortran/src/tensogram.f90 [hunk 19]
```diff
@@ -0,0 +1,1436 @@
...
+   !> `hash_c` (a caller-owned target buffer that must outlive the C call) and
+   !> `hash_ptr`: a present-but-empty name means "no hash" (NULL); absent means
+   !> the default "xxh3".
+   subroutine resolve_hash(hash, hash_c, hash_ptr)
+      character(len=*), intent(in), optional :: hash
+      character(kind=c_char), allocatable, target, intent(out) :: hash_c(:)
+      type(c_ptr),                                 intent(out) :: hash_ptr
+      character(len=:), allocatable :: hash_s
+      if (present(hash)) then
+         hash_s = hash
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.f90]: Fixed in f5ebf56 — `resolve_hash` now `trim`s the hash name, so a space-padded `character(len=N) :: hash = 'xxh3'` reaches C as `xxh3` (and an all-blank name correctly means no hash).

File: fortran/src/tensogram.f90 [hunk 20]
```diff
@@ -0,0 +1,1436 @@
...
+   end subroutine file_assign
+
+   !> Open an existing .tgm file for reading.
+   subroutine tensogram_file_open(path, file, err)
+      character(len=*),      intent(in)  :: path
+      type(tensogram_file),  intent(out) :: file
+      integer(c_int),        intent(out) :: err
+      character(kind=c_char), allocatable, target :: path_c(:)
+      type(c_ptr) :: out
+      call f_to_cstr(path, path_c)
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.f90]: Fixed in f5ebf56 — `tensogram_file_open` trims the path, so a padded `get_command_argument` variable opens the intended filename.

File: fortran/src/tensogram.f90 [hunk 21]
```diff
@@ -0,0 +1,1436 @@
...
+   end subroutine tensogram_file_open
+
+   !> Create a new .tgm file for writing (append messages with file_append).
+   subroutine tensogram_file_create(path, file, err)
+      character(len=*),      intent(in)  :: path
+      type(tensogram_file),  intent(out) :: file
+      integer(c_int),        intent(out) :: err
+      character(kind=c_char), allocatable, target :: path_c(:)
+      type(c_ptr) :: out
+      call f_to_cstr(path, path_c)
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.f90]: Fixed in f5ebf56 — `tensogram_file_create` trims the path as well; a test asserts a space-padded path creates the trimmed filename.

File: fortran/src/tensogram.f90 [hunk 22]
```diff
@@ -0,0 +1,1436 @@
...
+   !> descriptors map; objects are supplied via _write); defaults to `{}`.
+   subroutine tensogram_streaming_encoder_create(path, enc, err, metadata_json, hash)
+      character(len=*),                  intent(in)  :: path
+      type(tensogram_streaming_encoder), intent(out) :: enc
+      integer(c_int),                    intent(out) :: err
+      character(len=*), intent(in), optional :: metadata_json, hash
+      character(kind=c_char), allocatable, target :: path_c(:), meta_c(:), hash_c(:)
+      character(len=:),       allocatable         :: meta_s
+      type(c_ptr) :: hash_ptr, out
+      call f_to_cstr(path, path_c)
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/src/tensogram.f90]: Fixed in f5ebf56 — `tensogram_streaming_encoder_create` trims the path too. The three metadata getter keys were trimmed in the same commit for the same reason.

File: AGENTS.md [hunk 23]
```diff
@@ -197,6 +197,8 @@ Serves at http://localhost:8000/ (set BASE_PATH env var to deploy under a subpat
       `find . -name package.json -not -path './**/node_modules/*' -not -path './target/*'`
       (currently `typescript/` and `examples/typescript/`; the list grows if
       new JS packages land).
+    - `fortran/fpm.toml` `version` and the `project(... VERSION ...)` line in
+      `fortran/CMakeLists.txt` (the Fortran binding).
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply AGENTS.md:201]: Fixed in 9a3c969 — `release-preflight.yml` now validates `fortran/fpm.toml`'s `version` and the `project(... VERSION ...)` in `fortran/CMakeLists.txt` against the expected release version, so the VERSION single-source-of-truth list in AGENTS.md is actually enforced for the Fortran manifests.

File: plans/TEST.md [hunk 24]
```diff
@@ -20,6 +20,7 @@ Repo: ecmwf/tensogram
 | `tensogram-cli` | `src/**/*.rs` unit (per-subcommand) | CLI argument parsing + subcommand behaviour, incl. `--features netcdf,grib` for converter coverage |
 | `tensogram-ffi` | `src/lib.rs` unit (indirect via C++ wrapper) | FFI handle behaviour, null-pointer safety, error-code mapping |
 | C++ wrapper | `cpp/tests/*.cpp` (GoogleTest) | RAII handle behaviour, typed exception mapping, cross-language round-trip |
+| Fortran binding | `fortran/test/*.f90` + cross-lang drivers (`xlang_c.cpp`, `parity_*.py`), run via CTest | Generic dtype/rank round-trip, file API, metadata + lossless compression, non-copyable guard (negative), error-enum↔header consistency, **bidirectional** Fortran↔C/C++ parity, Fortran→Python parity |
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply plans/TEST.md]: Fixed in b34aec3 — the Fortran row now reads "bidirectional Fortran↔C/C++ and Fortran↔Python parity", matching the two CMake-registered parity tests (`parity_fortran_to_python` and `parity_python_to_fortran`).

File: fortran/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,171 @@
...
+  add_test(NAME test_copy_guard COMMAND test_copy_guard)
+  set_tests_properties(test_copy_guard PROPERTIES WILL_FAIL TRUE)
+
+  # Every handle type's non-copyable guard (and tensogram_check's failure
+  # path) must abort. One program, one WILL_FAIL test per mode.
+  add_executable(test_guards test/test_guards.f90)
+  target_link_libraries(test_guards PRIVATE tensogram_f)
+  foreach(mode buffer message file metadata stream check_ctx check_noctx)
+    add_test(NAME guard_${mode} COMMAND test_guards ${mode})
+    set_tests_properties(guard_${mode} PROPERTIES WILL_FAIL TRUE)
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply fortran/CMakeLists.txt]: Fixed in 5ea6b1b — replaced `WILL_FAIL` with a per-mode `PASS_REGULAR_EXPRESSION` matching each guard's specific abort message. CTest then checks the captured output (and ignores the non-zero `error stop` exit), so a setup failure that aborts *before* the guard — a different message — now fails the test instead of passing. Verified all 8 guard tests pass.

File: docs/src/guide/fortran-api.md [hunk 26]
```diff
@@ -0,0 +1,394 @@
...
+| `tensogram_streaming_encoder_count(enc)` / `_finish(enc, err)` / `enc%free()` | Objects written so far / finalise (footer + close) / release |
+| `buf%as_array(out)` / `buf%size()` / `buf%free()` | Buffer access and release |
+| `file%close()` | Close a file handle (also automatic at scope exit) |
+| `tensogram_check` / `tensogram_last_error` / `tensogram_strerror` | Error helpers |
+
+`tensogram_encode`, `tensogram_to_array`, `tensogram_file_append`, and
+`tensogram_streaming_encoder_write` are generic interfaces over **dtype**
+(`real32`, `real64`, `int32`, `int64`) and **rank** (`0`–`7`). `a` / `out` are assumed-rank; the dtype is
+resolved from the array's type/kind, the rank from the array itself. A
+dtype mismatch on decode returns `TGM_ERROR_OBJECT`.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply docs/src/guide/fortran-api.md]: Already addressed in 6d59fcd (this thread is on an outdated line). Line 252 no longer calls the dummies assumed-rank; it now reads "The dtype is resolved from the array's type/kind, the rank from the array itself (ranks above 7 are unsupported and fail to compile)", matching the explicit per-rank F2008 implementation.

File: plans/TODO.md [hunk 27]
```diff
@@ -165,6 +165,44 @@ streaming encoder.  See the *Asynchronous C++ API* entry in
...
 
+  - [ ] **fortran-interface — shallow `iso_c_binding` binding over the C FFI**
+    Let Fortran scientific codes (IFS and friends) read and write
+    Tensogram directly. **The full analysis and the staged roadmap live
+    in [`PLAN_FORTRAN.md`](../PLAN_FORTRAN.md) at the repo root — read it
+    before starting.** Decided shape:
+    - **Path A**: a hand-written `iso_c_binding` Fortran module
+      (`fortran/src/tensogram.F90`) binding straight to the existing
+      `libtensogram` C ABI — no new C or Rust code, no C++ detour. This
+      is the `eccodes_f90` model ECMWF Fortran users already expect.
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply plans/TODO.md]: Fixed in b34aec3 — the entry is rewritten as delivered: it now states the synchronous binding has shipped (library, tests, examples, docs, CI) and lists only the genuine remaining follow-ups (async surface, extra dtypes, optional `CFI_cdesc_t` zero-copy), pointing at PLAN_FORTRAN.md and the CHANGELOG.
